File: src/fe/src/models/Session.ts
/**
 * Session Model - Tracks individual transcription sessions (sermons)
 * Inspired by Glass's session management architecture
 */

export interface TranscriptSegment {
  id: string;
  text: string;
  timestamp: Date;
  confidence: number;
}

export interface ScriptureReference {
  id: string;
  book: string;
  chapter: number;
  verse: number;
  version: string;
  text: string;
  confidence: number;
  transcriptSegmentId: string;
}

export interface SessionMetadata {
  title: string;
  username: string;
  speaker?: string;
  location?: string;
  notes?: string;
}

export interface Checkpoint {
  id: string;
  timestamp: Date;
  audioPosition: number; // Seconds from start
  scriptureCount: number;
  scriptures: Array<{
    book: string;
    chapter: number;
    verse: number;
    reference: string;
    timestamp: number;
  }>;
  transcript: string;
  settings: {
    selectedModel: string;
    apiKey: string;
  };
}

export type SessionStatus = 'active' | 'completed' | 'error';

export interface Session {
  id: string;
  sessionCode: string; // Format: "a23h-username" (4-digit alphanumeric + username)
  userName: string;
  title: string;
  startedAt: Date;
  endedAt?: Date;
  updatedAt: Date;
  status: SessionStatus;
  isReadonly: boolean;
  isRecording?: boolean;
  isPaused?: boolean;
  
  // Duration tracking
  duration: number; // Total seconds
  activeDuration: number;
  
  // Content tracking
  transcripts: TranscriptSegment[];
  scriptureReferences: ScriptureReference[];
  scriptureCount: number;
  scriptures: Array<{
    book: string;
    chapter: number;
    verse: number;
    reference: string;
    timestamp: number;
  }>;
  transcript: string;
  
  // Metadata
  metadata: SessionMetadata;
  
  // Error recovery
  checkpoints: Checkpoint[];
  lastCheckpointAt?: Date;
  hasError: boolean;
  errorMessage?: string;
  
  // Settings snapshot (for recovery)
  settingsSnapshot: {
    selectedModel: string;
    apiKey: string;
  };
}

export interface SessionSummary {
  id: string;
  sessionCode: string;
  startedAt: Date;
  duration: number;
  scriptureCount: number;
  status: SessionStatus;
  title: string;
  username: string;
}

File: src/be/Models/Session.cs
namespace HOPTranscribe.Models;

public class Session
{
    public string Id { get; set; } = null!;
    public string SessionCode { get; set; } = null!;
    public string UserName { get; set; } = null!;
    public string Title { get; set; } = null!;
    public DateTime StartedAt { get; set; }
    public DateTime? EndedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public SessionStatus Status { get; set; }
    public bool IsReadonly { get; set; }
    public bool IsRecording { get; set; }
    public bool IsPaused { get; set; }
    
    // Duration tracking
    public int Duration { get; set; } // Total duration in seconds
    public int ActiveDuration { get; set; } // Seconds actually recording
    
    // Content
    public List<TranscriptSegment> Transcripts { get; set; } = new();
    public List<ScriptureReference> ScriptureReferences { get; set; } = new();
    
    // Metadata
    public SessionMetadata? Metadata { get; set; }
}

public class TranscriptSegment
{
    public string Id { get; set; } = null!;
    public string Text { get; set; } = null!;
    public DateTime Timestamp { get; set; }
    public double Confidence { get; set; }
}

public class ScriptureReference
{
    public string Id { get; set; } = null!;
    public string Book { get; set; } = null!;
    public int Chapter { get; set; }
    public int Verse { get; set; }
    public string Version { get; set; } = null!;
    public string Text { get; set; } = null!;
    public double Confidence { get; set; }
    public string TranscriptSegmentId { get; set; } = null!;
}

public class SessionMetadata
{
    public string? Speaker { get; set; }
    public string? Location { get; set; }
    public string? Notes { get; set; }
}

public enum SessionStatus
{
    Active,
    Completed,
    Ended,
    Error
}
